internal/shared/util/finalizer/finalizer.go

```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2025.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package finalizer
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"sort"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// EnsureFinalizer adds one or more finalizers to the object using server-side apply.
+// If all finalizers already exist, this is a no-op and returns (false, nil).
+// Returns (true, nil) if any finalizers were added.
+// Note: This function will update the passed object with the server response.
+func EnsureFinalizer(ctx context.Context, c client.Client, obj client.Object, finalizers ...string) (bool, error) {
+	if len(finalizers) == 0 {
+		return false, nil
+	}
+
+	// Determine which finalizers need to be added
+	var toAdd []string
+	for _, finalizer := range finalizers {
+		if !controllerutil.ContainsFinalizer(obj, finalizer) {
+			toAdd = append(toAdd, finalizer)
+		}
+	}
+
+	if len(toAdd) == 0 {
+		return false, nil
+	}
+
+	// Sort the finalizers to add for consistent ordering
+	sort.Strings(toAdd)
+
+	// Create a copy of the finalizers list to avoid mutating the object
+	currentFinalizers := obj.GetFinalizers()
+	newFinalizers := make([]string, len(currentFinalizers), len(currentFinalizers)+len(toAdd))
+	copy(newFinalizers, currentFinalizers)
+	newFinalizers = append(newFinalizers, toAdd...)
+
+	// Get the GVK for this object type
+	gvk, err := apiutil.GVKForObject(obj, c.Scheme())
+	if err != nil {
+		return false, fmt.Errorf("getting object kind: %w", err)
+	}
+
+	// Create an unstructured object for server-side apply
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(gvk)
+	u.SetName(obj.GetName())
+	u.SetNamespace(obj.GetNamespace())
+	u.SetFinalizers(newFinalizers)
+
+	// Use server-side apply to update finalizers
+	if err := c.Patch(ctx, u, client.Apply, client.ForceOwnership, client.FieldOwner("finalizer-controller")); err != nil {
```

Comment:
each controller (ClusterExtension, Catalog, ClusterExtensionRevisio/n) should has its own field owner, following this convention:

```
olm.operatorframework.io/<CONTROLLER-NAME>
```

for example: 
* `olm.operatorframework.io/clusterextension-controller`
* `olm.operatorframework.io/clustercatalog-controller`
* `olm.operatorframework.io/clusterextensionrevision-controller`

Reply:
I would I would suggest that because the finalizers are being treated "separately" that they have a different owner referencing "finalizer".
That being said, this is generic code shared among all three... I'll have to add another argument.